bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+		NotRoutable,
+		Transport,
+		DestinationUnsupported,
+		ExceedsMaxMessageSize,
+		MissingArgument,
+		Fees,
+	}
+
+	impl<T: Config> From<XcmpSendError> for Error<T> {
+		fn from(e: XcmpSendError) -> Self {
+			match e {
+				XcmpSendError::NotApplicable => Error::<T>::Send(SendError::NotApplicable),
+				XcmpSendError::Unroutable => Error::<T>::Send(SendError::NotRoutable),
+				XcmpSendError::Transport(_) => Error::<T>::Send(SendError::Transport),
+				XcmpSendError::DestinationUnsupported =>
+					Error::<T>::Send(SendError::DestinationUnsupported),
+				XcmpSendError::ExceedsMaxMessageSize =>
+					Error::<T>::Send(SendError::ExceedsMaxMessageSize),
+				XcmpSendError::MissingArgument => Error::<T>::Send(SendError::MissingArgument),
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	/// The nonce of the message been processed or not
+	#[pallet::storage]
+	pub type NoncesBitmap<T: Config> = StorageMap<_, Twox64Concat, u128, u128, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit an inbound message originating from the Gateway contract on Ethereum
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit())]
+		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
+			let _who = ensure_signed(origin)?;
+			ensure!(!Self::operating_mode().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&message.event_log, &message.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			// Decode event log into an Envelope
+			let envelope =
+				Envelope::try_from(&message.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Verify the message has not been processed
+			ensure!(!<Nonce<T>>::get(envelope.nonce.into()), Error::<T>::InvalidNonce);
+
+			// Decode payload into `MessageV2`
+			let message = MessageV2::decode_all(&mut envelope.payload.as_ref())
+				.map_err(|_| Error::<T>::InvalidPayload)?;
+
+			let xcm =
+				T::MessageConverter::convert(message).map_err(|e| Error::<T>::ConvertMessage(e))?;
+
+			// Todo: Deposit fee(in Ether) to RewardLeger which should cover all of:
+			// T::RewardLeger::deposit(who, envelope.fee.into())?;
+			// a. The submit extrinsic cost on BH
+			// b. The delivery cost to AH
+			// c. The execution cost on AH
+			// d. The execution cost on destination chain(if any)
+			// e. The reward
+
+			// Attempt to forward XCM to AH
+			let dest = Location::new(1, [Parachain(T::AssetHubParaId::get())]);
+			let (message_id, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
```

Comment:
@vgeddes I added the SendController. I encountered two issues:
- [pallet-xcm adds a DescendOrigin](https://github.com/claravanstaden/polkadot-sdk/blob/master/polkadot/xcm/pallet-xcm/src/lib.rs#L2446) when sending the XCM with the [relayer account as origin](https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-bfe761b371a553619edf5be2dd792c45bd14f1a916e51b77f5b77b9395d92cdaR271). The fee teleport then fails on AH because the relayer account is not a trusted teleporter. I am not sure how to fix this. I think previously we didn't have this problem because the message origin was BH sovereign, which is allowed to teleport.
- pallet-xcm on BH does not allow sending arbitrary XCM. [I changed this](https://github.com/claravanstaden/polkadot-sdk/pull/4/files#diff-bdd3c9c10b342ec9c25583cec417f2b1ad7eaca27046c2f4a41651b1a8d200c9R253), not sure if it breaks some security model or if it is allowed.

Reply:
Ah too bad. Looks like SendController is too high-level for our needs. 

Lets stick with the existing approach then

Reply:
Reverted in https://github.com/claravanstaden/polkadot-sdk/pull/4/commits/89746964578648edd2e9eebea83ad39b065d4d01.

Reply:
If we want to charge delivery fee from the relayer,  one option is to use `ExecuteController` instead.

```
Withdraw(TeleportFee)
PayFees(TeleportFee)
InitiateTeleport(dest: AH, xcm: xcm)// xcm is just the current we build
```